docs/plugin/interfaces/guidelines.md

```diff
@@ -0,0 +1,383 @@
+
+* When an interface needs to be extended instead of adding new methods to the existing one, one could add the extended functionality to a new interface. Now the client can also work with components that do not yet implement the new extended functionality (as it can find out with QueryInterface if it is available) while on the other hand you are not forcing an update of all the components that do implement the initial interface, you only update the one(s) for which the extended new functionality makes sense.
+
+Design interfaces to be asynchronous when needed {#AsynchronousInterfaces}
+------------------------------------------------
+
+Sometimes when designing an interface it can be expected that the implementation of a certain action can be quite expensive and take some time. 
+
+In such a case (we usually use an expected duration of a few hundred milliseconds as the limit) or when the duration of an action is non deterministic and worse case would take longer than a few hundred milliseconds, it is advised to make the interface asynchronous. The interface will then only pass the data needed to start the action to the component and add a callback to the interface to make it possible for the implementation to return the result asynchronously. The component will handle the action in the background (do not use threads for this but use the Thunder workerpool) and call the provided callback when the action is completed to signal this to the initiator of the action. Not only will this prevent the call to fail because of JSON-RPC or COM-RPC timeouts but it will also make sure the framework is not blocked internally and also the client can continue without being blocked waiting for the result (or needed to reserve a thread for this).
+
+For example an interface that will require user interaction for its action (like asking for the pin to be entered) is a perfect example of one that should be made asynchronous as it is of course completely nondeterministic when the call will return as the successful entering of the PIN could take minutes/hours potentially.
+
+Also note, as was pointed out in the "State complete interfaces" paragraph of this page, that there should also be a means to stop the initiated action before it has completed. This is for example needed when the component that started the action needs to shut down and with this it can signal that the action can be stopped as it is no longer needed and the provided callback is no longer valid and should not longer be called.
+
+
```

Comment:
Do we need guidance on choice of JSON over COM when exposing an interface?

Reply:
Hi @lucienkennedylamb see if I understand your remark correctly: you mean when as a 'customer of the interface' you should use one over the other? Or when as an interface provider decide if to provide both or only COM-RPC?

Reply:
I was alluding to the interface developer - choosing to provide both or only COM-RPC, so I'm wondering if we should provide guidance on how to make that decision.
But actually you're right about the 'customer' or client of an interface too - If they have both options available to them should they always go with COM-RPC over JSON-RPC which provides stronger typing?

Reply:
Thanks for clearing this up, I'll add some text to describe at least the options and when you should consider what. 
On your question for the customer of the interface our advise would be to use COM-RPC for the native world, it is strong typed and binary (so will reduce resource usage). JSON-RPC is targeted at the webworld where that is much easier to use.